User Streams API Support

kolinkrewinkel opened the pull request:
This technically works, but needs to be reviewed. Line 148 specifically. After the pause/setValue/resume exchange, the queue behaves strangely for at least one operation (not starting it.)

Commit:
Added SocketShuttle

Commit:
Updated SocketShuttle

kolinkrewinkel committed:
Fixes non-buildability.

kolinkrewinkel committed:
Merge branch 'master' of github.com:joeldev/ADNKit into user-streams

kolinkrewinkel committed:
Fix spacing

kolinkrewinkel committed:
Completes stream context objects and conform with Joel's fucking coding style.

kolinkrewinkel committed:
Begin outlining stream protocols.

kolinkrewinkel committed:
Revise naming of request method.

kolinkrewinkel committed:
Add method in implementation body.

kolinkrewinkel committed:
Conform to coding style and add a semaphore.

kolinkrewinkel committed:
Begin attemps to connect.

kolinkrewinkel committed:
Make sure I\'m not missing anything.

kolinkrewinkel committed:
Remove frameworks from the list.

kolinkrewinkel committed:
Spawns request and at least gets data back.

kolinkrewinkel committed:
Switch to a set and add some basic handling code.

kolinkrewinkel committed:
Connects and disconnects. Warns of the current issue; SocketShuttle also does not remove itself as an observer: will need to be fixed.

kolinkrewinkel committed:
Switch away from using the easily-broken property-delegate method; passes method; passes it around instead. Now we can just filter with anyObject for one with a conn. ID and no socket shuttle and go, when connecting.

kolinkrewinkel committed:
Switch context objects to store delegates instead of blocks.

kolinkrewinkel committed:
Call back the delegate.

kolinkrewinkel committed:
Maybe we're getting somewhere...

kolinkrewinkel committed:
There, this fixes things and prevents hold ups and queue bullshit.

kolinkrewinkel committed:
Fixes header and removes unneeded properties and initializers.

kolinkrewinkel committed:
Hey, it actually works now.

kolinkrewinkel committed:
Minor revisions to structuring and verbiage.

kolinkrewinkel committed:
Switch to just passing back the entirety of the JSON.

kolinkrewinkel committed:
Okay, now it actually messages the delegate. Yes.

kolinkrewinkel committed:
Minor revisions.

kolinkrewinkel committed:
Okay, so we\'ll have to identify the incoming JSON\'s type to parse it.

kolinkrewinkel committed:
This should add parsing.

kolinkrewinkel committed:
Remove debug log.

kolinkrewinkel committed:
Minor optimization.

kolinkrewinkel committed:
Implement all of the new KATSocketShuttleDelegate methods and send ones for connecting/opening.

kolinkrewinkel committed:
Add convenience method.

kolinkrewinkel committed:
Revise to match project coding style.

kolinkrewinkel committed:
Add some convenience methods for getting the right delegates and generally improving how things are dispatched.

kolinkrewinkel committed:
Get your types right, idiot.

kolinkrewinkel committed:
Concurrency = no.

kolinkrewinkel committed:
Adds a convenience method and sends the rest of the methods.

kolinkrewinkel committed:
>_>

kolinkrewinkel committed:
More minor cleanup

kolinkrewinkel committed:
Transfers over the completion block. Kinda critical, and really it should just be copied.

kolinkrewinkel committed:
This is ghetto, but it will work.

kolinkrewinkel committed:
Leftovers or something.

kolinkrewinkel committed:
Add stream contexts to combination-AFNetworking target.

kolinkrewinkel committed:
Add SocketShuttle to the AFNetworking-hybrid target.

kolinkrewinkel committed:
Fixes the issue with freezing queues (but with a hacky approach) and cleans the code a little. Needs further review.

kolinkrewinkel committed:
Merge master into user-streams.

kolinkrewinkel committed:
Fixes non-building status.

kolinkrewinkel commented:
This can be merged now, per-approval (including the hacky fix on line 148.) I'd prefer said hack fixed before it gets merged though, if it's possible.

kolinkrewinkel committed:
Code cleanup.

kolinkrewinkel committed:
Fix a compiler warning over ANKOEmbed... which was out of nowhere.

kolinkrewinkel committed:
Fix kv coding error. Test things first kolin

kolinkrewinkel committed:
...And again with the KV coding.

kolinkrewinkel committed:
Switch to use the constant.

kolinkrewinkel committed:
Hmm...this might actually be fixed.

kolinkrewinkel committed:
...Fixes the never-resuming operation.

kolinkrewinkel committed:
This adds reconnect support. Both SocketShuttle and SocketRocket have been updated on my local machine though... but both need major updates (and ARC) as they're riddled with leaks. Subsequently, on reconnects the SRWebSocket calls a dangling pointer and causes crashes.

kolinkrewinkel committed:
..Maybe this is the trick?

kolinkrewinkel committed:
Begin adding reachability-connection-type flags

kolinkrewinkel committed:
Fixes non-building status with AFNetworking-bundled target.

kolinkrewinkel committed:
Prevents operations from stalling out forever.

kolinkrewinkel committed:
...now it works. removed the dispatch_async stuff in the socketshuttle. it could've been fixed by running the reconfigure block in a dispatch_async, but it would've made the double indirection not work.

tonyarnold commented:
What's the hack, @kolinkrewinkel? I'm about to start an app build that needs user streams, so I'd be happy to have a look.

tonyarnold committed:
Use the correct developer frameworks variable so that builds work properly under Xcode 5

joeldev commented:
Sorry @kolinkrewinkel, I think I'm going to have to reject this for a few reasons. Firstly, the ShocketShuttle submodule should be added using the https repo URL instead of the SSH URL. Secondly, the project doesn't build the OS X framework target successfully, which is a deal-killer. Can you try taking another look at these issues? Once these issues are resolved I'd be happy to approve it. Thanks!

joeldev commented:
Looks like @tonyarnold is working on these same issues as well.

kolinkrewinkel commented:
Psst Joel you're supposed to leave it open

tonyarnold committed:
Update SocketShuttle submodule to use HTTPS URL

tonyarnold commented:
I'm going to see where I get this afternoon — I had a few hours to start a new app idea, but I can't start without OS X and user streams support. I'll submit a new PR referencing this one against the user-streams branch when it's ready.

kolinkrewinkel commented:
:), thanks @joeldev. @tonyarnold, I may have resolved it, but it requires further review. It was with operations in the NSOperationQueue (-operationQueue) not starting after pausing the streaming operation, even after it was resumed. Querying it would reveal that they were simply lying there with the state "isReady." Recently it hasn't been giving me trouble, but it was such a weird quirk I'm sure it's something obvious I'm just not seeing.

A new set of issues is with SocketShuttle and SocketRocket. We need to roll our own forks (we already have one for SocketShuttle) for two reasons. One is that when the SocketShuttle is recreated, it appears to leave the SocketRocket instance alive for a time (or indefinitely as a leak—see point 2) which causes a message to be sent to a deallocated SocketShuttle instance. ARC would fix this just as a start (with weak references). Two is that just from the fact that the SRWebSocket instance isn't deallocating immediately, something is up with it internally (either it's keeping itself alive with retain cycles or failing to close, which I believe is a prerequisite to it deallocating.)

kolinkrewinkel commented:
Also @tonyarnold, feel free to ping me with a private message (on ADN) and we can talk real-time and collaborate on it, if you want. Also available on AIM as kkrewink.

tonyarnold committed:
Re-enable OS X support

tonyarnold committed:
Update submodule reference to latest SocketShuttle changes that include OS X support

tonyarnold committed:
Make missing project headers public in the framework

tonyarnold commented:
If you're planning on forking SocketRocket, check my fork for a couple of really minor project fixes.

kolinkrewinkel committed:
Merge pull request #58 from tonyarnold/user-streams

Re-enable OS X support for the user-streams branch

tternes commented:
Any plans to fork/merge @tonyarnold's changes for SocketRocket? I had a to make a few changes to get it to build debug on 10.8. I also have a few changes to ADNKit and SocketShuttle to get OS X working that I can get a PR put together for.

kolinkrewinkel commented:
We don't actually have a repository for that (and that's not my area.) I'll work on it more when I have time... first day of sophomore year went from "free coding day" to "don't show me Objective-C; I need sleep."

tternes commented:
No problem. I'll commit my changes to a fork and sort out a pull request for @joeldev to consider.

keichan34 commented:
Any word on when this will be merged into master?

I can confirm this works well on OS X (using @tternes fork, with [one minor change in SocketShuttle](https://github.com/keichan34/SocketShuttle/commit/d3dadea517da3903f6af034259ec0daf171aba61) to get the dependencies to install).

mthurman committed:
Allow multiple streaming operations to be started

When trying to subscribe to multiple UserStream topics, only the
first one was succeeding (since it got paused and then resumed on
didReceiveMessage in the isInitialConnectionResponse state). This
pauses subsequent operations that come in while we're trying to
connect the websocket so once we successfully connect, we can
resume (and subscribe with the correct connection_id).

mthurman committed:
Add meta params applicable to user streams

berg committed:
Merge pull request #71 from mthurman/multiple-user-streams

Allow multiple streaming operations to be started

berg committed:
Merge pull request #72 from mthurman/user-streams-meta

Add meta params applicable to user streams

berg commented:
What do you guys think about getting this merged to master?

@kolinkrewinkel, @mthurman and I hacked around on this last weekend (well, really just those two guys), and it seems to work well. I'm going to try to bring this up to date so that it's based on top of master, and then we can see how it goes.

kolinkrewinkel commented:
As soon as it's brought up to date, it can be merged.

Sent from my iPad

> On Jan 30, 2014, at 17:32, Bryan Berg notifications@github.com wrote:
> 
> What do you guys think about getting this merged to master?
> 
> @kolinkrewinkel, @mthurman and I hacked around on this last weekend (well, really just those two guys), and it seems to work well. I'm going to try to bring this up to date so that it's based on top of master, and then we can see how it goes.
> 
> —
> Reply to this email directly or view it on GitHub.

kolinkrewinkel commented:
Does anyone know how we can go about adding a dependency to joeldev/SocketShuttle in the pod spec? I'm not well versed in pod-knowledge and AFAIK you can't do git references within a pod spec. It's the only hold back before merging.

tonyarnold commented:
You'd have to create a new podspec and submit it — something like ADNKitSocketShuttle. What's holding the changes back from being merged upstream?

kolinkrewinkel commented:
We've never actually talked about merging them back up. I guess there's nothing wrong with doing so, though.

tonyarnold commented:
I dunno, it just seems like good form to offer to push them back upstream and then use that as your dependency.

berg commented:
Good call, @tonyarnold. My first step is going to be to take all of the substantive changes from the SocketShuttle fork, make them backwards compatible, and (try to) get them accepted upstream.

keichan34 commented:
[This commit](https://github.com/joeldev/SocketShuttle/pull/9) doesn't need to be merged, because they already fixed that issue upstream, I think.

berg commented:
OK, sent in a pull request for SocketShuttle as a start: https://github.com/mk/SocketShuttle/pull/3

It's basically all of the meat of Kolin's SocketShuttle patches, without the packaging changes, which I think we can accommodate inside of the ADNKit project without requiring a custom fork of SocketShuttle. This should make it easier to keep the projects manageable both by CocoaPods and git submodule users.

One thing I did run into here, which @kolinkrewinkel might have already addressed, but I haven't yet found: in my code I copy ANKClient a lot, specifically because I want to make requests with the same access token but with different pagination parameters. We should probably ensure that this doesn't mean we're creating new sockets every time (I think it's a fairly common pattern.)

kolinkrewinkel commented:
Couple things: firstly, thanks. My initial inclination is towards a singleton socket, or—even better—one that sticks per authentication token.

Now that we're in the conclusion phase, I'm curious what my rationale was to destroying and recreating the socket shuttle instances on disconnect or error. In my opinion this defeats the purpose of using SocketShuttle over SocketRocket (among other things.) So, before merging, ideally, I'd like to investigate...not doing that.

Sent from my iPhone

> On Feb 17, 2014, at 10:42, Bryan Berg notifications@github.com wrote:
> 
> OK, sent in a pull request for SocketShuttle as a start: mk/SocketShuttle#3
> 
> It's basically all of the meat of Kolin's SocketShuttle patches, without the packaging changes, which I think we can accommodate inside of the ADNKit project without requiring a custom fork of SocketShuttle. This should make it easier to keep the projects manageable both by CocoaPods and git submodule users.
> 
> One thing I did run into here, which @kolinkrewinkel might have already addressed, but I haven't yet found: in my code I copy ANKClient a lot, specifically because I want to make requests with the same access token but with different pagination parameters. We should probably ensure that this doesn't mean we're creating new sockets every time (I think it's a fairly common pattern.)
> 
> —
> Reply to this email directly or view it on GitHub.

tonyarnold commented:
Yeah, one per token would probably be best — multi-account support would be a bit hard with a singleton. Maybe a singleton socket handler that vends socket instances per client? Would you ever need to establish more than one socket per app per client?

kolinkrewinkel commented:
Not that I can think of. Even changing pagination settings wouldn't warrant creating a new socket as it's only for the short-lifetime of the request.

berg commented:
I was just going to suggest creating a singleton containing a weak NSHashMap of access_token -> SocketShuttle instances and having the ANKClient lazily create (and retain) them when needed, nilling the reference when the accessToken property changes.

tonyarnold commented:
You're just looking for an excuse to use `NSHashTable`, @berg :trollface:

tonyarnold commented:
Has this been brought up to date as per @berg's earlier comments? I'm debating dropping this in in place of my own, manual AFNetworking/Socket-based code in a project.

_Update: it might be up-to-date, but it doesn't compile out of the box on OS X. SocketRocket/SocketShuttle keep trying to build the iOS static libs_ :cry:

tonyarnold committed:
ADNKit-OSX had the iOS SocketShuttle as a target dependency, fixed.

tonyarnold committed:
NSString+ANKAdditions header should be public

kolinkrewinkel committed:
Merge pull request #75 from tonyarnold/user-streams

ADNKit-OSX had the iOS SocketShuttle as a target dependency, fixed.

kolinkrewinkel commented:
So, remaining, (and feel free to suggest, so I can add to this):
- [ ] Test and hopefully amend destruction and recreation of `socketShuttle` property.
- [ ] Change to global per-token pattern of sockets, rather than per-client.
- [ ] Have SocketShuttle PR merged.
- [ ] Test fresh installs with both submodules and CocoaPods.
- [ ] Documentation
- [ ] Merge! ✨

tonyarnold commented:
Silly question, but how do I subscribe to a user's stream? Maybe add:
- [ ] Documentation

_Update: Sorted this out privately with @kolinkrewinkel — I was on the right track, but misunderstood ADN's recommendations when starting a stream_

tonyarnold committed:
Add support for the inactive field on Channel objects

tonyarnold commented:
In addition to the global per-token sockets, it'd be good to have a status indicator for whether the current client has an open/closed/errored socket — or some method of stopping double streaming subscriptions.

berg committed:
Merge pull request #78 from tonyarnold/user-streams-inactive-field

Add support for the inactive field on Channel objects

garrettmurray commented:
Any update on this? Is it stable enough to use? I'm working with ADNKit now and I'd love to use user streams from the get-go.

kolinkrewinkel commented:
You should be able to use this branch and stream right from the start. We just have the revisions in the checklist to do for (what we guess are) potential bugs and semantics.

tonyarnold commented:
Don't hate me, but I ran into some problems I couldn't be bothered diagnosing with AFN 1.x and started working on AFN 2.x support: https://github.com/tonyarnold/ADNKit/tree/afnetworking-2

It's mostly functional, and I think it'll end up being a bit simpler to implement/look after. It does not support anything less than iOS 7.x or OS X 10.9 though, so probably not anyone else's cup of tea at this stage :tea:

tonyarnold commented:
Something we need to add to the streaming request: configurable parameters. I just lost a morning trying to work out why I wasn't getting any annotations through with my streamed messages.

keichan34 commented:
@tonyarnold I'm using `-(ANKGeneralParameters *) generalParameters` on the main `ANKClient` object, is that not working for you?

tonyarnold commented:
Not in the streamed messages, no. No parameters are currently passed to the initial request for the stream, so annotations, etc aren't included. I've just pushed an update to my (getting out of control) AFNetworking 2 branch that includes this functionality.

keichan34 commented:
Hm, that's strange. I set `generalParameters` to the appropriate values (in my case, annotations and no HTML), log in with the access token (or request one if necessary), then initialize streaming. Works fine for me... (FWIW, I'm using 9aae7dd738f39a090dbfd8ea1ff27f954caef9f4 at the moment)

tonyarnold commented:
You see annotations in your streamed responses? I don't see how — they're not included by default, and the current user-streams branch doesn't ask for them… can you verify that you're seeing any annotations in the actual streamed responses (not the initial response you get when registering from a stream)?

keichan34 commented:
Updating to the latest commit on `user-streams` seemed to make the annotations go away (and the HTML to come back). I can't see anything in the commit logs that indicate things have changed, so maybe it's an updated dependency?

tonyarnold commented:
Possibly, although local testing showed that if I didn't make that initial request to `https://stream-channel.app.net/stream/user` with the `include_annotations=1` query parameter included, no annotations come through. Maybe it's an ADN bug/change?
